src/operator/nn/pooling.cc

```diff
@@ -98,9 +103,24 @@ static bool PoolingShape(const nnvm::NodeAttrs &attrs,
       << "Pooling: Input data should be  3D in (batch, channel, x)"
       << " Or 4D in (batch, channel, y, x) "
       << " Or 5D in (batch, channel, d, y, x)";
+  CHECK_LE(dshape.ndim(), 5U)
+      << "Pooling: Input data should be  3D in (batch, channel, x)"
+      << " Or 4D in (batch, channel, y, x) "
+      << " Or 5D in (batch, channel, d, y, x)";
   TShape oshape = dshape;
   if (dshape.ndim() == 0) return false;
-  if (param.kernel.ndim() == 1) {
+  if (param.global_pool) {
+      if (dshape.ndim() == 3) {
+          oshape[2] = 1;
+      } else if (dshape.ndim() == 4) {
+          oshape[2] = 1;
+          oshape[3] = 1;
+      } else if (dshape.ndim() == 5) {
+          oshape[2] = 1;
+          oshape[3] = 1;
+          oshape[4] = 1;
+      }
```

Comment:
Need to push the oshape to `out_shape`. See https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/pooling.cc#L163-L168.

Reply:
Also, you can use a for-loop instead of if.

Reply:
There is another mismatch at this line:
https://github.com/apache/incubator-mxnet/blob/c9ec3118688c233a66ad847003a9e8d2d09e5952/src/operator/nn/pool.h#L690
and I have no idea how I could fix this as this function pool() have no global_pool as input parameter.

Reply:
I feel it is sort of dangerous to modify function definitions and add a input parameter. Do we have other choices ?

Reply:
There is no need to revise this function. All you need to do is to prepare the correct initial values. You can check the logic here https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/pooling-inl.h#L135-L149 . If global_pool is used, kernel is set to be `TShape(ishape.data()+ishape.ndim()-param_.kernel.ndim(), ishape.data()+ishape.ndim())`, pad is set to be all 0 and stride is set to be all 1. You can write a function to do this. For example:
```c++
TShape kernel = param_.kernel;
TShape padding = param_.pad;
TShape stride = param_.stride;
if(param_.global_pool) {
   kernel = TShape(ishape.data()+ishape.ndim()-param_.kernel.ndim(), ishape.data()+ishape.ndim());
   padding = TShape(ishape.ndim() - 2);
   for(int i = 0; i < ishape.ndim() - 2; i++) {
      padding[i] = 0;
   }
   stride = TShape(ishape.ndim() - 2);
}
pool(s, in_data.dptr<DType>(), in_data.shape_, out_data.shape_,
     kernel,
     padding,
     stride,
     param_.pool_type, req, out_data.dptr<DType>());
```